File: Example/Example/ViewController.m
#import "ViewController.h"
#import <ObjectiveDDP/BSONIdGenerator.h>

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UITableView *tableview;
@property (copy, nonatomic) NSString *listName;

@end

@implementation ViewController

- (id)initWithNibName:(NSString *)nibNameOrNil
               bundle:(NSBundle *)nibBundleOrNil
               meteor:(MeteorClient *)meteor
             listName:(NSString *) listName {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        self.meteor = meteor;
        self.listName = listName;
    }
    return self;
}

- (void)viewWillAppear:(BOOL)animated {
    self.navigationItem.title = self.listName;
    UIBarButtonItem *addButton = [[UIBarButtonItem alloc ] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd
                                                                                target:self
                                                                                action:@selector(didTouchAdd:)];
    [self.navigationItem setRightBarButtonItem:addButton];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didReceiveUpdate:)
                                                 name:@"added"
                                               object:nil];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didReceiveUpdate:)
                                                 name:@"removed"
                                               object:nil];
}

- (void)viewWillDisappear:(BOOL)animated {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)didReceiveUpdate:(NSNotification *)notification {
    [self.tableview reloadData];
}

- (NSArray *)computedList {
    NSPredicate *pred = [NSPredicate predicateWithFormat:@"(listName like %@)", self.listName];
    return [self.meteor.collections[@"things"] filteredArrayUsingPredicate:pred];
}

#pragma mark UI Actions

- (IBAction)didTouchAdd:(id)sender {
    AddViewController *addController = [[AddViewController alloc] initWithNibName:@"AddViewController"
                                                                           bundle:nil];
    addController.delegate = self;
    [self presentViewController:addController animated:YES completion:nil];
}

#pragma mark <UITableViewDataSource>

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.computedList count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *cellIdentifier = @"thing";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellIdentifier];

    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault
                                      reuseIdentifier:cellIdentifier];
    }

    NSDictionary *thing = self.computedList[indexPath.row];
    cell.textLabel.text = thing[@"msg"];

    return cell;
}

- (BOOL)tableView:(UITableView *)tableView
        canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (void)tableView:(UITableView *)tableView
        commitEditingStyle:(UITableViewCellEditingStyle)editingStyle
        forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        NSDictionary *thing = self.computedList[indexPath.row];
        [self.meteor callMethodName:@"/things/remove" parameters:@[@{@"_id": thing[@"_id"]}] responseCallback:nil];
    }
}

- (void)didAddThing:(NSString *)message {
    [self dismissViewControllerAnimated:YES completion:nil];
    NSArray *parameters = @[@{@"_id": [[NSUUID UUID] UUIDString],
                              @"msg": message,
                              @"owner": self.userId,
                              @"listName": self.listName}];
    [self.meteor callMethodName:@"/things/insert" parameters:parameters responseCallback:nil];
}

@end

File: Example/swiftExample/swiftddp/ViewController.swift
//
//  ViewController.swift
//  swiftddp
//
//  Created by Michael Arthur on 7/6/14.
//  Copyright (c) 2014. All rights reserved.
//

import UIKit

class ViewController: UIViewController,UITableViewDataSource, AddViewControllerDelegate {
    
    var meteor:MeteorClient!
    var listName:String!
    var userId:String!
    
    @IBOutlet weak var tableview: UITableView!
    
    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
        
        
    }
    
    override init(nibName nibNameOrNil: String!, bundle nibBundleOrNil: NSBundle!) {
        
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        
    }
    
    init(nibNameOrNil: String!, bundle nibBundleOrNil: NSBundle!, meteor: MeteorClient!, listName:String!) {
        
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        // if(self != nil) {
        self.meteor = meteor
        self.listName = listName
        //}
    }
    
    override func viewWillAppear(animated: Bool) {
        self.navigationItem.title = self.listName
        let addButton:UIBarButtonItem = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.Add, target: self, action: "didTouchAdd:")
        
        self.navigationItem.setRightBarButtonItem(addButton, animated: true)

        NSNotificationCenter.defaultCenter().addObserver(self, selector: "didReceiveUpdate:", name: "things_added", object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "didReceiveUpdate:", name: "things_changed", object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "didReceiveUpdate:", name: "things_removed", object: nil)
    }
    
    override func viewWillDisappear(animated: Bool) {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
    
    
    func didReceiveUpdate(notification:NSNotification) {
        self.tableview.reloadData()
    }
    
    func computedList() -> NSArray {
        
        
        let pred:NSPredicate = NSPredicate(format: "(listName like %@)", self.listName)
        let temp = self.meteor.collections["things"] as! M13MutableOrderedDictionary
        let temp2 = temp.allObjects() as NSArray
        return temp2.filteredArrayUsingPredicate(pred)
        
        
    }
    
    @IBAction func didTouchAdd(sender: AnyObject) {
        let addController = AddViewController(nibName: "AddViewController", bundle: nil)
        
        addController.delegate = self
        self.presentViewController(addController, animated: true, completion: nil)
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if(self.meteor.collections["things"] != nil){
            
            return self.computedList().count
        }
        return 0
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cellIdentifier:String! = "thing"
        var cell:UITableViewCell
        
        if let tmpCell: AnyObject = tableView.dequeueReusableCellWithIdentifier(cellIdentifier) {
            cell = tmpCell as! UITableViewCell
        } else {
            cell = UITableViewCell(style: UITableViewCellStyle.Default, reuseIdentifier: cellIdentifier) as UITableViewCell
        }
        
        if(self.meteor.collections["things"] != nil){
            let thing:NSDictionary = self.computedList()[indexPath.row] as! NSDictionary
            cell.textLabel?.text = thing["msg"] as? String
            return cell
        }
        cell.textLabel?.text = "dummy"
        return cell
        
    }
    
    func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {
        return true
    }
    
    func tableView(tableView: UITableView, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath) {
        if(editingStyle == UITableViewCellEditingStyle.Delete) {
            //If statement prevents crash
            if(self.meteor.collections["things"] != nil){
                let thing:NSDictionary = self.computedList()[indexPath.row] as! NSDictionary
                let thingy = thing["_id"] as! String
                self.meteor.callMethodName("/things/remove", parameters: [["_id":thingy]], responseCallback: nil)
            }
        }
    }
    
    func didAddThing(message: NSString!) {
        self.dismissViewControllerAnimated(true, completion: nil)
        let parameters:NSArray = [["_id": NSUUID().UUIDString,
            "msg":message,
            "owner":self.userId,
            "listName":self.listName]]
        
        self.meteor.callMethodName("/things/insert", parameters: parameters as [AnyObject], responseCallback: nil)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
}

